modules/http/scanner.go

```diff
@@ -158,12 +158,40 @@ func (scan *scan) withDeadlineContext(ctx context.Context) context.Context {
 
 // Dial a connection using the configured timeouts, as well as the global deadline, and on success,
 // add the connection to the list of connections to be cleaned up.
-func (scan *scan) dialContext(ctx context.Context, net string, addr string) (net.Conn, error) {
+func (scan *scan) dialContext(ctx context.Context, network string, addr string) (net.Conn, error) {
 	dialer := zgrab2.GetTimeoutConnectionDialer(scan.scanner.config.Timeout)
 
+	switch network {
+	case "tcp", "tcp4", "tcp6", "udp", "udp4", "udp6":
+		// If the scan is for a specific IP, and a domain name is provided, we
+		// don't want to just let the http library resolve the domain.  Create
+		// a fake resolver that we will use, that always returns the IP we are
+		// given to scan.
+		if scan.target.IP != nil && scan.target.Domain != "" {
+			host, _, err := net.SplitHostPort(addr)
+			if err != nil {
+				log.Errorf("http/scanner.go dialContext: unable to split host:port '%s'", addr)
+				log.Errorf("No fake resolver, IP address may be incorrect: %s", err)
+			} else {
+				// In the case of redirects, we don't want to blindly use the
+				// IP we were given to scan, however.  Only use the fake
+				// resolver if the domain originally specified for the scan
+				// target matches the current address being looked up in this
+				// DialContext.
+				if host == scan.target.Domain {
+					resolver, err := zgrab2.NewFakeResolver(scan.target.IP.String())
+					if err != nil {
+						return nil, err
+					}
+					dialer.Dialer.Resolver = resolver
+				}
+			}
+		}
+	}
+
```

Comment:
This hunk (lines 164-181) addresses the concerns @justinbastress raised, and also adds an additional check against the specified `network` (similar to what `internetAddrList` does in the go net library)

Reply:
I think I follow -- this `dialContext` gets called by the HTTP lib for all redirects, and it can decide to pass in the `addr` as a `hostname(?!):port`; if the redirect is to our "pinned" host, set a custom resolver for this connection only (a subsequent redirect leads to another `dialContext`, another `Dialer`, etc). If it's to another host, we use the default resolver (but if that host redirects back to our pinned host, we are again by another call to this).

Reply:
Yep!